ethexe/sequencer/src/lib.rs

```diff
@@ -18,212 +18,183 @@
+            .block_end_state_is_valid(hash)
+            .is_none_or(|valid| !valid)
+        {
+            bail!("Block {hash} database state is not valid");
+        }
 
-        if let Some(candidate) = codes_candidate {
-            code_commitments_len = candidate.commitments().len() as u64;
-            log::debug!("Collected {code_commitments_len} code commitments. Trying to submit...");
+        // TODO: add status
+        self.waiting_for_commitments = self
+            .db
+            .block_commitment_queue(hash)
+            .ok_or_else(|| anyhow!("Block {hash} has not block commitment queue"))?
+            .into_iter()
+            .collect();
 
-            codes_future = Some(self.submit_codes_commitment(candidate));
-        };
+        // Remove all commitments, which we are not waiting for anymore.
+        self.block_commitments
+            .retain(|_, c| self.waiting_for_commitments.contains(&c.commitment.hash));
 
-        if let Some(candidate) = blocks_candidate {
-            block_commitments_len = candidate.commitments().len() as u64;
-            log::debug!("Collected {block_commitments_len} block commitments. Trying to submit...",);
+        self.blocks_candidate.take();
+        self.codes_candidate.take();
 
-            blocks_future = Some(self.submit_block_commitments(candidate));
-        };
+        log::debug!("[SEQUENCER] Collection round for {hash} started");
+        self.collection_round.start(hash);
 
-        match (codes_future, blocks_future) {
-            (Some(codes_future), Some(transitions_future)) => {
-                let (codes_tx, transitions_tx) = futures::join!(codes_future, transitions_future);
-                codes_tx?;
-                transitions_tx?;
-            }
-            (Some(codes_future), None) => codes_future.await?,
-            (None, Some(transitions_future)) => transitions_future.await?,
-            (None, None) => {}
+        if let Some(block) = self.validation_round.stop() {
+            log::debug!("[SEQUENCER] Validation round for {block} stopped");
         }
 
-        self.update_status(|status| {
-            status.submitted_code_commitments += code_commitments_len;
-            status.submitted_block_commitments += block_commitments_len;
-        });
-
         Ok(())
     }
 
+    pub fn get_candidate_code_commitments(&self) -> impl Iterator<Item = &CodeCommitment> + '_ {
+        Self::get_candidate_commitments(&self.codes_candidate, &self.code_commitments)
+    }
+
+    pub fn get_candidate_block_commitments(&self) -> impl Iterator<Item = &BlockCommitment> + '_ {
+        Self::get_candidate_commitments(&self.blocks_candidate, &self.block_commitments)
+    }
+
     pub fn receive_code_commitments(
         &mut self,
         aggregated: AggregatedCommitments<CodeCommitment>,
     ) -> Result<()> {
-        log::debug!("Received code commitments: {aggregated:?}");
+        log::debug!("[SEQUENCER] Received code commitments: {aggregated:?}");
```

Comment:
```suggestion
        log::debug!("Received code commitments: {aggregated:?}");
```

Logger will write module before message